### PR TITLE
Add caching bad token detector

### PR DESCRIPTION
**Pull request by vkgnosis**
### Test Plan
new tests

### CODE REVIEW

File: orderbook/src/bad_token/cache.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+use super::{BadTokenDetecting, TokenQuality};
+use anyhow::Result;
+use primitive_types::H160;
+use std::{collections::HashMap, sync::Mutex};
+
+pub struct CachingDetector {
+    inner: Box<dyn BadTokenDetecting>,
+    // std mutex is fine because we don't hold lock across await.
+    cache: Mutex<HashMap<H160, TokenQuality>>,
```
Comment (bh2smith) [inline orderbook/src/bad_token/cache.rs:9]: Ahh! Now I see how/where the token address is associated with its quality.

**Review by bh2smith — APPROVED** [review]

File: orderbook/src/bad_token/cache.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+#[async_trait::async_trait]
+impl BadTokenDetecting for CachingDetector {
+    async fn detect(&self, token: H160) -> Result<TokenQuality> {
+        if let Some(result) = self.get_from_cache(&token) {
```
Comment (fleupold) [inline orderbook/src/bad_token/cache.rs:15]: Would it make sense to have a ttl for cache entries? I imaging some tokens can change their state (e.g. getting paused or a fee becoming enabled) so maybe caching results for 7 days on prod before we restart is a bit long?

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia3.giphy.com%2Fmedia%2FRgfGmnVvt8Pfy%2Fgiphy.gif&token=nice" data-gitmeme-token="nice"></a>

File: orderbook/src/bad_token/cache.rs (hunk 2 above)
Comment (vkgnosis) [reply orderbook/src/bad_token/cache.rs:15]: Good idea. Will remember this in an issue.

**Pull request merged**